Add ParseStrict and MustParseStrict methods.

mattmoyer opened the pull request:
These are methods like `Parse` and `MustParse`, but validate that their inputs contain only valid base32 characters. On invalid input, they return a new `ErrInvalidCharacters` error rather than producing an undefined ULID.

A comparison benchmark:
```
$ go test -v -run=^$ -benchtime 10s -bench BenchmarkParse ./...
goos: darwin
goarch: amd64
pkg: github.com/oklog/ulid
BenchmarkParse-8         	500000000	        26.1 ns/op	 994.49 MB/s
BenchmarkParseStrict-8   	500000000	        39.2 ns/op	 662.87 MB/s
PASS
ok  	github.com/oklog/ulid	39.221s
?   	github.com/oklog/ulid/cmd/ulid	[no test files]
```

This is an alternate fix for https://github.com/oklog/ulid/issues/21. I originally approached this issue by adding this validation to `Parse` and `UnmarshalText` (https://github.com/oklog/ulid/pull/22), but there was some concern about the performance impact.

Fixes https://github.com/oklog/ulid/issues/21.

mattmoyer commented:
One question about the interface here as long as we're adding a new method: should this require that the input is _exactly_ the same format returned by `.String()`?

For example, it currently accepts lowercase base32. Should it reject those?

mattmoyer commented on AUTHORS.md:
```diff
@@ -1,2 +1,3 @@
 - Peter Bourgon (@peterbourgon)
 - Tomás Senart (@tsenart)
+- Matt Moyer (@mattmoyer)
```
Is this expected? Happy to drop this line or leave it.

tsenart replied:
Yeah, this is capturing original authors. For contributors we rely on Github: https://github.com/oklog/ulid/graphs/contributors

tsenart commented:
> For example, it currently accepts lowercase base32. Should it reject those?

I don't think it should reject those.

tsenart commented on ulid.go:
```diff
@@ -276,6 +291,55 @@ func (id *ULID) UnmarshalText(v []byte) error {
+//
+// ErrDataSize is returned if the len(v) is different from an encoded
+// ULID's length. Invalid encodings return ErrInvalidCharacters.
+func (id *ULID) UnmarshalTextStrict(v []byte) error {
```
The reason for the naming `UnmarshalText` is that it implements the [text.Unmarshaler](https://golang.org/pkg/encoding/#TextUnmarshaler) interface. I don't see a reason to have this symmetry here.

mattmoyer replied:
Makes sense, thanks.

tsenart commented on ulid.go:
```diff
@@ -276,6 +291,55 @@ func (id *ULID) UnmarshalText(v []byte) error {
+		dec[v[25]] == 0xFF {
+		return ErrInvalidCharacters
+	}
+	return id.UnmarshalText(v)
```
`UnmarshalText` already checks for encoded size as well as overflow. With this implementation we'd be repeating those checks twice.

mattmoyer replied:
Makes sense. I was originally thinking about preserving the errors returned for length/`v[0]` overflow, but on second thought it doesn't really matter that these new methods return those errors exactly the same.

tsenart replied:
How about factoring out those common checks into a private function that can be reused by both `Parse` and `ParseStrict`?

tsenart replied:
> That's a little tricky because Parse() calls through to UnmarshalText() which also needs to do the length check.

Then we can take out the code from `UnmarshalText` so that it can be reused.

> There's definitely an opportunity for a small optimization here, though. The other thing is that the ErrOverflow check can be omitted after the ErrInvalidCharacters check has been done, since the v[0] > '7' condition is a subset of ErrInvalidCharacters.

Nice.

mattmoyer commented:
@tsenart: thanks for the review!

I updated to address review comments, please take a look at your leisure. The only other new bit I added was `MustParseStrict()` which is just like `MustParse()` but with `ParseStrict()` underneath.

mattmoyer commented on ulid_test.go:
```diff
@@ -157,6 +168,36 @@ func TestMarshalingErrors(t *testing.T) {
+
+	type testCase struct {
+		name  string
+		input string
```
Ah, yeah this was a holdover from the previous error behavior. Updated to check against _exactly_ `ErrInvalidCharacters` which is the expected error in all cases.

mattmoyer commented:
@tsenart I updated to split out some internal validation functions that are shared between the parsing functions:

- `validateSize()`
- `validateOverflow()`
- `validateCharacterSet()`
- `decodeInto()`

Unfortunately, some of these functions are too big for Go to want to inline them, which hurts performance:
```
benchmark                  old ns/op     new ns/op     delta
BenchmarkParse-8           25.6          30.4          +18.75%
BenchmarkParseStrict-8     35.7          44.2          +23.81%

benchmark                  old MB/s     new MB/s     speedup
BenchmarkParse-8           1014.82      854.11       0.84x
BenchmarkParseStrict-8     728.90       588.59       0.81x
```

In addition to the function call overhead, I think there are some bounds checks that can't be elided because the code isn't inlined.

The only other approach I can think of would be to copy-paste the code and manually inline it at both locations, which feels messy to me.

tsenart commented:
We're dealing here directly with the trade-off of performance for maintainability and usability. @peterbourgon: I know you said you didn't want to make a judgment call on this, but I'd like your thoughts on this again.

peterbourgon commented:
You can get the existing API with the old (better) perf numbers by removing some function calls.

```diff
commit 86816e50b7e734ddd361c3fc332e041f034ce57e
Author: Peter Bourgon <peter@bourgon.org>
Date:   Mon Jul 16 10:09:12 2018 -0700

    De-function the ParseStrict branch

diff --git a/ulid.go b/ulid.go
index adb826e..a857395 100644
--- a/ulid.go
+++ b/ulid.go
@@ -101,7 +101,7 @@ func MustNew(ms uint64, entropy io.Reader) ULID {
 // ULID's length. Invalid encodings produce undefined ULIDs. For a version that
 // returns an error instead, see ParseStrict.
 func Parse(ulid string) (id ULID, err error) {
-	return id, id.UnmarshalText([]byte(ulid))
+	return id, parse([]byte(ulid), false, &id)
 }
 
 // ParseStrict parses an encoded ULID, returning an error in case of failure.
@@ -112,18 +112,81 @@ func Parse(ulid string) (id ULID, err error) {
 // ErrDataSize is returned if the len(ulid) is different from an encoded
 // ULID's length. Invalid encodings return ErrInvalidCharacters.
 func ParseStrict(ulid string) (id ULID, err error) {
-	v := []byte(ulid)
-	if err := validateSize(v); err != nil {
-		return id, err
+	return id, parse([]byte(ulid), true, &id)
+}
+
+func parse(v []byte, strict bool, id *ULID) error {
+	// Check if if a base32 encoded ULID is the right length.
+	if len(v) != EncodedSize {
+		return ErrDataSize
 	}
-	if err := validateCharacterSet(v); err != nil {
-		return id, err
+
+	// Check if all the characters in a base32 encoded ULID are part of the
+	// expected base32 character set.
+	//
+	// Important: must only be called with len(v) == EncodedSize.
+	if strict &&
+		(dec[v[0]] == 0xFF ||
+			dec[v[1]] == 0xFF ||
+			dec[v[2]] == 0xFF ||
+			dec[v[3]] == 0xFF ||
+			dec[v[4]] == 0xFF ||
+			dec[v[5]] == 0xFF ||
+			dec[v[6]] == 0xFF ||
+			dec[v[7]] == 0xFF ||
+			dec[v[8]] == 0xFF ||
+			dec[v[9]] == 0xFF ||
+			dec[v[10]] == 0xFF ||
+			dec[v[11]] == 0xFF ||
+			dec[v[12]] == 0xFF ||
+			dec[v[13]] == 0xFF ||
+			dec[v[14]] == 0xFF ||
+			dec[v[15]] == 0xFF ||
+			dec[v[16]] == 0xFF ||
+			dec[v[17]] == 0xFF ||
+			dec[v[18]] == 0xFF ||
+			dec[v[19]] == 0xFF ||
+			dec[v[20]] == 0xFF ||
+			dec[v[21]] == 0xFF ||
+			dec[v[22]] == 0xFF ||
+			dec[v[23]] == 0xFF ||
+			dec[v[24]] == 0xFF ||
+			dec[v[25]] == 0xFF) {
+		return ErrInvalidCharacters
 	}
-	if err := validateOverflow(v); err != nil {
-		return id, err
+
+	// Check if the first character in a base32 encoded ULID will overflow. This
+	// happens because the base32 representation encodes 130 bits, while the
+	// ULID is only 128 bits.
+	//
+	// See https://github.com/oklog/ulid/issues/9 for details.
+	//
+	// Important: must only be done after size has been validated.
+	if v[0] > '7' {
+		return ErrOverflow
 	}
-	decodeInto(v, &id)
-	return id, nil
+
+	// 6 bytes timestamp (48 bits)
+	(*id)[0] = ((dec[v[0]] << 5) | dec[v[1]])
+	(*id)[1] = ((dec[v[2]] << 3) | (dec[v[3]] >> 2))
+	(*id)[2] = ((dec[v[3]] << 6) | (dec[v[4]] << 1) | (dec[v[5]] >> 4))
+	(*id)[3] = ((dec[v[5]] << 4) | (dec[v[6]] >> 1))
+	(*id)[4] = ((dec[v[6]] << 7) | (dec[v[7]] << 2) | (dec[v[8]] >> 3))
+	(*id)[5] = ((dec[v[8]] << 5) | dec[v[9]])
+
+	// 10 bytes of entropy (80 bits)
+	(*id)[6] = ((dec[v[10]] << 3) | (dec[v[11]] >> 2))
+	(*id)[7] = ((dec[v[11]] << 6) | (dec[v[12]] << 1) | (dec[v[13]] >> 4))
+	(*id)[8] = ((dec[v[13]] << 4) | (dec[v[14]] >> 1))
+	(*id)[9] = ((dec[v[14]] << 7) | (dec[v[15]] << 2) | (dec[v[16]] >> 3))
+	(*id)[10] = ((dec[v[16]] << 5) | dec[v[17]])
+	(*id)[11] = ((dec[v[18]] << 3) | dec[v[19]]>>2)
+	(*id)[12] = ((dec[v[19]] << 6) | (dec[v[20]] << 1) | (dec[v[21]] >> 4))
+	(*id)[13] = ((dec[v[21]] << 4) | (dec[v[22]] >> 1))
+	(*id)[14] = ((dec[v[22]] << 7) | (dec[v[23]] << 2) | (dec[v[24]] >> 3))
+	(*id)[15] = ((dec[v[24]] << 5) | dec[v[25]])
+
+	return nil
 }
 
 // MustParse is a convenience function equivalent to Parse that panics on failure
@@ -278,14 +341,7 @@ const EncodedSize = 26
 // ErrDataSize is returned if the len(v) is different from an encoded
 // ULID's length. Invalid encodings produce undefined ULIDs.
 func (id *ULID) UnmarshalText(v []byte) error {
-	if err := validateSize(v); err != nil {
-		return err
-	}
-	if err := validateOverflow(v); err != nil {
-		return err
-	}
-	decodeInto(v, id)
-	return nil
+	return parse(v, false, id)
 }
 
 // Time returns the Unix time in milliseconds encoded in the ULID.
@@ -389,89 +445,3 @@ func (id *ULID) Scan(src interface{}) error {
 func (id ULID) Value() (driver.Value, error) {
 	return id.MarshalBinary()
 }
-
-// validateSize is an internal function to check if a base32 encoded ULID is
-// the right length.
-func validateSize(v []byte) error {
-	if len(v) != EncodedSize {
-		return ErrDataSize
-	}
-	return nil
-}
-
-// validateOverflow is an internal function to check if the first character in
-// a base32 encoded ULID will overflow. This happens because the base32
-// representation encodes 130 bits, while the ULID is only 128 bits.
-//
-// See https://github.com/oklog/ulid/issues/9 for details.
-//
-// Important: must only be called after size has been validated.
-func validateOverflow(v []byte) error {
-	if v[0] > '7' {
-		return ErrOverflow
-	}
-	return nil
-}
-
-// validateCharacterSet is an internal function to check if all the characters
-// in a base32 encoded ULID are part of the expected base32 character set.
-//
-// Important: must only be called with len(v) == EncodedSize.
-func validateCharacterSet(v []byte) error {
-	if dec[v[0]] == 0xFF ||
-		dec[v[1]] == 0xFF ||
-		dec[v[2]] == 0xFF ||
-		dec[v[3]] == 0xFF ||
-		dec[v[4]] == 0xFF ||
-		dec[v[5]] == 0xFF ||
-		dec[v[6]] == 0xFF ||
-		dec[v[7]] == 0xFF ||
-		dec[v[8]] == 0xFF ||
-		dec[v[9]] == 0xFF ||
-		dec[v[10]] == 0xFF ||
-		dec[v[11]] == 0xFF ||
-		dec[v[12]] == 0xFF ||
-		dec[v[13]] == 0xFF ||
-		dec[v[14]] == 0xFF ||
-		dec[v[15]] == 0xFF ||
-		dec[v[16]] == 0xFF ||
-		dec[v[17]] == 0xFF ||
-		dec[v[18]] == 0xFF ||
-		dec[v[19]] == 0xFF ||
-		dec[v[20]] == 0xFF ||
-		dec[v[21]] == 0xFF ||
-		dec[v[22]] == 0xFF ||
-		dec[v[23]] == 0xFF ||
-		dec[v[24]] == 0xFF ||
-		dec[v[25]] == 0xFF {
-		return ErrInvalidCharacters
-	}
-	return nil
-}
-
-// decodeInto is an internal function that uses an optimized unrolled loop
-// (borrowed from https://github.com/RobThree/NUlid) to decode a base32 ULID
-// into a *ULID destination.
-//
-// Important: must only be called with len(v) == EncodedSize.
-func decodeInto(v []byte, id *ULID) {
-	// 6 bytes timestamp (48 bits)
-	(*id)[0] = ((dec[v[0]] << 5) | dec[v[1]])
-	(*id)[1] = ((dec[v[2]] << 3) | (dec[v[3]] >> 2))
-	(*id)[2] = ((dec[v[3]] << 6) | (dec[v[4]] << 1) | (dec[v[5]] >> 4))
-	(*id)[3] = ((dec[v[5]] << 4) | (dec[v[6]] >> 1))
-	(*id)[4] = ((dec[v[6]] << 7) | (dec[v[7]] << 2) | (dec[v[8]] >> 3))
-	(*id)[5] = ((dec[v[8]] << 5) | dec[v[9]])
-
-	// 10 bytes of entropy (80 bits)
-	(*id)[6] = ((dec[v[10]] << 3) | (dec[v[11]] >> 2))
-	(*id)[7] = ((dec[v[11]] << 6) | (dec[v[12]] << 1) | (dec[v[13]] >> 4))
-	(*id)[8] = ((dec[v[13]] << 4) | (dec[v[14]] >> 1))
-	(*id)[9] = ((dec[v[14]] << 7) | (dec[v[15]] << 2) | (dec[v[16]] >> 3))
-	(*id)[10] = ((dec[v[16]] << 5) | dec[v[17]])
-	(*id)[11] = ((dec[v[18]] << 3) | dec[v[19]]>>2)
-	(*id)[12] = ((dec[v[19]] << 6) | (dec[v[20]] << 1) | (dec[v[21]] >> 4))
-	(*id)[13] = ((dec[v[21]] << 4) | (dec[v[22]] >> 1))
-	(*id)[14] = ((dec[v[22]] << 7) | (dec[v[23]] << 2) | (dec[v[24]] >> 3))
-	(*id)[15] = ((dec[v[24]] << 5) | dec[v[25]])
-}
```

benchstat on master to this branch:

```
bork ~ benchstat master.txt mattmoyer.txt
name         old time/op    new time/op    delta
Parse-4        26.7ns ± 2%    28.9ns ± 2%  +8.33%  (p=0.008 n=5+5)
MustParse-4    31.2ns ± 3%    33.4ns ± 1%  +7.25%  (p=0.008 n=5+5)

name         old speed      new speed      delta
Parse-4       975MB/s ± 2%   901MB/s ± 2%  -7.64%  (p=0.008 n=5+5)
MustParse-4   834MB/s ± 3%   778MB/s ± 1%  -6.77%  (p=0.008 n=5+5)

name         old alloc/op   new alloc/op   delta
Parse-4         0.00B          0.00B         ~     (all equal)
MustParse-4     0.00B          0.00B         ~     (all equal)

name         old allocs/op  new allocs/op  delta
Parse-4          0.00           0.00         ~     (all equal)
MustParse-4      0.00           0.00         ~     (all equal)
```

benchstat of this branch to my proposed changes:

```
bork ~ benchstat mattmoyer.txt peterbourgon.txt
name           old time/op    new time/op    delta
Parse-4          28.9ns ± 2%    26.9ns ± 3%   -6.86%  (p=0.008 n=5+5)
ParseStrict-4    40.9ns ± 5%    36.2ns ± 3%  -11.31%  (p=0.008 n=5+5)
MustParse-4      33.4ns ± 1%    31.3ns ± 5%   -6.52%  (p=0.008 n=5+5)

name           old speed      new speed      delta
Parse-4         901MB/s ± 2%   967MB/s ± 3%   +7.35%  (p=0.008 n=5+5)
ParseStrict-4   637MB/s ± 5%   718MB/s ± 3%  +12.67%  (p=0.008 n=5+5)
MustParse-4     778MB/s ± 1%   833MB/s ± 5%   +7.08%  (p=0.008 n=5+5)

name           old alloc/op   new alloc/op   delta
Parse-4           0.00B          0.00B          ~     (all equal)
ParseStrict-4     0.00B          0.00B          ~     (all equal)
MustParse-4       0.00B          0.00B          ~     (all equal)

name           old allocs/op  new allocs/op  delta
Parse-4            0.00           0.00          ~     (all equal)
ParseStrict-4      0.00           0.00          ~     (all equal)
MustParse-4        0.00           0.00          ~     (all equal)
```

benchstat of master to my proposed changes:

```
bork ~ benchstat master.txt peterbourgon.txt
name         old time/op    new time/op    delta
Parse-4        26.7ns ± 2%    26.9ns ± 3%   ~     (p=0.722 n=5+5)
MustParse-4    31.2ns ± 3%    31.3ns ± 5%   ~     (p=0.730 n=5+5)

name         old speed      new speed      delta
Parse-4       975MB/s ± 2%   967MB/s ± 3%   ~     (p=0.841 n=5+5)
MustParse-4   834MB/s ± 3%   833MB/s ± 5%   ~     (p=0.841 n=5+5)

name         old alloc/op   new alloc/op   delta
Parse-4         0.00B          0.00B        ~     (all equal)
MustParse-4     0.00B          0.00B        ~     (all equal)

name         old allocs/op  new allocs/op  delta
Parse-4          0.00           0.00        ~     (all equal)
MustParse-4      0.00           0.00        ~     (all equal)
```

mattmoyer commented:
@peterbourgon that makes sense to me, will rebase this branch with that version shortly.

peterbourgon commented:
@tsenart I'm +1 on this API and implementation. How about you?

tsenart commented on ulid.go:
```diff
@@ -94,9 +98,94 @@ func MustNew(ms uint64, entropy io.Reader) ULID {
+	// ULID is only 128 bits.
+	//
+	// See https://github.com/oklog/ulid/issues/9 for details.
+	if v[0] > '7' {
```
We can save this check on the strict case since it’d be already caught by the above.

tsenart replied:
Apart from this, all looks good.

mattmoyer replied:
I thought so too, but a string like `99999999999999999999999999` passes the invalid character check but still overflows that first digit.

mattmoyer replied:
We could combine the checks in `ParseStrict` so `ErrInvalidCharacters` is returned in both cases, but I think it's more clear to get the explicit `ErrOverflow` error code for that special case.

tsenart commented:
CI isn’t working for external PRs currently because Travis doesn’t inject environment variables on those builds and the Coveralls reporting fails. We gotta make that step optional. Meanwhile we should merge locally.

nickform commented:
Just out of interest, when will there be a new release incorporating the new methods?

nickform commented:
v1 - 'grats on that milestone!

Thanks again for maintaining this library and to @mattmoyer for the initiative to support stronger validation.
